core/src/broadcast_stage/broadcast_utils.rs

```diff
@@ -26,45 +30,63 @@ pub struct UnfinishedSlotInfo {
     let timer = Duration::new(1, 0);
     let recv_start = Instant::now();
+
     let (mut bank, (entry, mut last_tick_height)) = receiver.recv_timeout(timer)?;
 
+    serialized_batch_byte_count += serialized_size(&entry)?;
     let mut entries = vec![entry];
-    let mut slot = bank.slot();
-    let mut max_tick_height = bank.max_tick_height();
-
-    assert!(last_tick_height <= max_tick_height);
-
-    if last_tick_height != max_tick_height {
-        while let Ok((try_bank, (entry, tick_height))) = receiver.try_recv() {
-            // If the bank changed, that implies the previous slot was interrupted and we do not have to
-            // broadcast its entries.
-            if try_bank.slot() != slot {
-                warn!("Broadcast for slot: {} interrupted", bank.slot());
-                entries.clear();
-                bank = try_bank;
-                slot = bank.slot();
-                max_tick_height = bank.max_tick_height();
-            }
-            last_tick_height = tick_height;
-            entries.push(entry);
-
-            if entries.len() >= RECEIVE_ENTRY_COUNT_THRESHOLD {
-                break;
-            }
-
-            assert!(last_tick_height <= max_tick_height);
-            if last_tick_height == max_tick_height {
-                break;
-            }
+
+    assert!(last_tick_height <= bank.max_tick_height());
+
+    // Drain channel
+    while last_tick_height != bank.max_tick_height() {
+        let (try_bank, (entry, tick_height)) = match receiver.try_recv() {
+            Ok(working_bank_entry) => working_bank_entry,
+            Err(_) => break,
+        };
+        // If the bank changed, that implies the previous slot was interrupted and we do not have to
+        // broadcast its entries.
+        if try_bank.slot() != bank.slot() {
+            warn!("Broadcast for slot: {} interrupted", bank.slot());
+            entries.clear();
+            serialized_batch_byte_count = 8; // Vec len
+            bank = try_bank;
+        }
+        last_tick_height = tick_height;
+        let entry_bytes = serialized_size(&entry)?;
+        serialized_batch_byte_count += entry_bytes;
+        entries.push(entry);
+        assert!(last_tick_height <= bank.max_tick_height());
+    }
+
+    // Wait up to `ENTRY_COALESCE_DURATION` to try to coalesce entries into a 32 shred batch
+    let coalesce_deadline = Instant::now() + ENTRY_COALESCE_DURATION;
+    while last_tick_height != bank.max_tick_height()
+        && serialized_batch_byte_count < target_serialized_batch_byte_count
+    {
+        let (try_bank, (entry, tick_height)) = match receiver.recv_deadline(coalesce_deadline) {
+            Ok(working_bank_entry) => working_bank_entry,
+            Err(_) => break,
+        };
+        // If the bank changed, that implies the previous slot was interrupted and we do not have to
+        // broadcast its entries.
+        if try_bank.slot() != bank.slot() {
```

Comment:
Don't we want to reset `coalesce_deadline` here as well?
```rust
coalesce_deadline = Instant::now() + ENTRY_COALESCE_DURATION;
```
Previous entries are thrown out, so the time it took to receive them is not  counted in broadcast of current entries.